src/sys/signal.rs

```diff
@@ -407,6 +426,108 @@ impl SigAction {
     }
 }
 
+/// Examine and change a signal action.
+///
+/// This function sets a signal action and returns the previous one. Parameters are:
+///
+/// * `signal` - the signal to modify.
+/// * `sigaction` - the new signal action.
+///
+/// Note that the `SIGKILL` and `SIGSTOP` signals are special and cannot be modified
+/// with this function.
+///
+/// Custom signal handlers should call only
+/// [signal-safe](https://wiki.sei.cmu.edu/confluence/display/c/SIG30-C.+Call+only+asynchronous-safe+functions+within+signal+handlers)
+/// functions. You should consult with a list of signal-safe functions for your system. For example,
+/// [POSIX](http://pubs.opengroup.org/onlinepubs/9699919799/functions/V2_chap02.html#tag_15_04_03_03)
+/// and [Linux](http://man7.org/linux/man-pages/man7/signal-safety.7.html) lists
+/// may serve as a good reference. Great care should be taken if the signal handler needs access
+/// to a global state. Global state used in signal handler must not be protected with
+/// [`Mutex`](https://doc.rust-lang.org/std/sync/struct.Mutex.html), it is better to stick to
+/// [`atomic`](https://doc.rust-lang.org/std/sync/atomic/index.html)s module or to use
+/// platform-specific tools which `nix` provides (like [pipe](../../unistd/pipe.v.html)).
+///
+/// For more information see the [`sigaction` man
+/// pages](http://pubs.opengroup.org/onlinepubs/9699919799/functions/sigaction.html).
+///
+/// # Safety 
+///
+/// This function is unsafe because it changes the global state of a program (signal handler
+/// itself). Custom signal handlers should follow the [rules for signal
+/// handlers](https://wiki.sei.cmu.edu/confluence/pages/viewpage.action?pageId=87152469).
+///
+/// # Examples
+///
+/// Here is an example of building asynchronous-safe signal handler. We're building a function that
+/// waits until `SIGINT` arrives and reports it. The problem is that reporting right from the
+/// signal handler may be unsafe if reporting code calls allocating or locking functions. This might
+/// be a problem if you're using `println!` or `sdtdout()` because there is a `lock` inside of
+/// them. So, one of the possible solutions is decoupling signal reporting and signal reaction.
+/// Signal handler should report to a signal reaction handler via write to a pipe.
+///
+/// ```no_run
+/// #[macro_use] extern crate lazy_static;
+/// extern crate libc;
+/// extern crate nix;
+/// 
+/// use libc::c_int;
+/// use std::os::unix::io::RawFd;
+/// 
+/// // The global state to be used in signal handler and function which will implement signal
+/// // reaction logic.
+/// static mut SIGNAL_FD: RawFd = -1;
+/// // We're not using Mutex<RawFd>, because calling lock/unlock in signal handler
+/// // is an unsafe operation, but we need a lock to set up SIGNAL_FD.
+/// lazy_static! {
+///     static ref SIGNAL_SET_LOCK: std::sync::Mutex<()> = std::sync::Mutex::new(());
```

Comment:
If you don't take the lock when reading SIGNAL_FD, then there's no point in taking the lock when writing it.  I advise deleting the mutex.